### PR TITLE
[INDDEX] Add Gaps By Item report

### PR DESCRIPTION
**Pull request by esoergel**
https://docs.google.com/document/d/1FeOQ63GaZyOeU1USjBgWhLn9wSBfOQoUiCo1lj5X51U/edit#

This is the last new report for INDDEX.  There may be some later churn once users start testing things out, but this is the final piece of the puzzle.

### CODE REVIEW

File: custom/inddex/reports/gaps_detail.py [hunk 1]
```diff
@@ -44,16 +46,27 @@ def __init__(self, food_data):
 
     @property
     def rows(self):
-        return []
+        rows = {}
+        for row in self._food_data.rows:
+            if row.conv_factor_gap_code != ConvFactorGaps.AVAILABLE:
+                key = (row.food_name, row.conv_method_code, row.conv_factor_gap_code)
```
Comment (esoergel) [inline custom/inddex/reports/gaps_detail.py:56]: IMO this doesn't make a lot of sense, as it only considers conversion factor gaps, but that's what was specified, so we'll start like this.
:shrug:

File: custom/inddex/reports/gaps_detail.py [hunk 2]
```diff
@@ -89,4 +93,13 @@ def __init__(self, food_data):
...
+        for row in _sort_rows(self._food_data.rows):
+            for gap_class, gap_code in [
+                    (ConvFactorGaps, row.conv_factor_gap_code),
+                    (FctGaps, row.fct_gap_code),
+            ]:
+                if gap_code != gap_class.AVAILABLE:
+                    manually_set = ['gap_type', 'gap_code', 'gap_desc']
+                    yield format_row([gap_class.name, gap_code, gap_class.get_description(gap_code)] + [
+                        getattr(row, col) for col in self.headers if col not in manually_set
+                    ])
```
Comment (esoergel) [inline custom/inddex/reports/gaps_detail.py:105]: This block here will have 0, 1, or 2 rows for each row in `_food_data`, depending on what gaps are present.

File: custom/inddex/tests/test_master_report.py [hunk 3]
```diff
@@ -224,8 +224,9 @@ def to_string(row):
         expected_report = get_expected_report(csv_filename)
         self.assertEqual(list(expected_report[0].keys()), actual_report.headers)
 
-        for expected_row, actual in zip(expected_report, actual_report.rows):
-            expected = list(expected_row.values())
+        expected_rows = sorted(list(r.values()) for r in expected_report)
+        actual_rows = sorted(actual_report.rows)
+        for expected, actual in zip(expected_rows, actual_rows):
```
Comment (esoergel) [inline custom/inddex/tests/test_master_report.py:229]: I was originally hoping to do all the sorting in the report, but the only columns known to be unique are case ID columns, which are random, so the generated report won't always match the test data.  This sorts in the test after swapping in the proper IDs.

**Review by kaapstorm — APPROVED** [review]
:tada:

**Pull request merged**